src/components/HybridAppMiddleware.tsx

```diff
@@ -1,104 +1,127 @@
+
+    // Save `exitTo` when we reach /transition route.
+    // `exitTo` should always exist during OldDot -> NewDot transitions.
+    useEffect(() => {
+        if (!NativeModules.HybridAppModule || !routeParams?.exitTo || exitTo) {
+            return;
         }
 
-        Navigation.navigate(exitUrl);
-    }, []);
+        Log.info('[HybridApp] Saving `exitTo` for later', true, {exitTo: routeParams?.exitTo});
+        setExitTo(routeParams?.exitTo);
 
-    /**
-     * This function only affects iOS. If during a single app lifecycle we frequently transition from OldDot to NewDot,
-     * we need to artificially show the bootsplash because the app is only booted once.
-     */
+        Log.info(`[HybridApp] Started transition`, true);
+        setStartedTransition(true);
+    }, [exitTo, routeParams?.email, routeParams?.exitTo]);
+
+    // This function only affects iOS. If during a single app lifecycle we frequently transition from OldDot to NewDot,
+    // we need to artificially show the bootsplash because the app is only booted once.
     const showSplashScreenOnNextStart = useCallback(() => {
+        Log.info('[HybridApp] Resetting the state of HybridAppMiddleware to show the BootSplash on the next transition', true);
         setIsSplashHidden(false);
         setStartedTransition(false);
         setFinishedTransition(false);
+        setExitTo(undefined);
     }, [setIsSplashHidden]);
 
     useEffect(() => {
-        if (!finishedTransition || isSplashHidden) {
+        if (!startedTransition || finishedTransition) {
             return;
         }
 
-        Log.info('[HybridApp] Finished transtion', true);
-        BootSplash.hide().then(() => {
-            setIsSplashHidden(true);
-            Log.info('[HybridApp] Handling onboarding flow', true);
-            Welcome.handleHybridAppOnboarding();
-        });
-    }, [finishedTransition, isSplashHidden, setIsSplashHidden]);
+        const transitionURL = NativeModules.HybridAppModule ? `${CONST.DEEPLINK_BASE_URL}${initialURL ?? ''}` : initialURL;
+        const isLoggingInAsNewUser = SessionUtils.isLoggingInAsNewUser(transitionURL ?? undefined, sessionEmail);
 
-    useEffect(() => {
-        if (!startedTransition) {
+        // We need to wait with navigating to exitTo until all login-related actions are complete.
+        if (!authenticated || isLoggingInAsNewUser || isAccountLoading) {
             return;
         }
 
-        // On iOS, the transitionEnd event doesn't trigger some times. As such, we need to set a timeout.
-        const timeout = setTimeout(() => {
-            setFinishedTransition(true);
-        }, CONST.SCREEN_TRANSITION_END_TIMEOUT);
+        if (exitTo) {
+            Navigation.isNavigationReady().then(() => {
+                // We need to remove /transition from route history.
+                // `useTransitionRouteParams` returns undefined for routes other than /transition.
+                if (routeParams) {
```

Comment:
I think this condition is unnecessary - `exitTo` is parsed from `routeParams` therefore it should always be truthy in this block 😄

Reply:
Sometimes we do navigation in different component. For public `/transition` route we can do it [here](https://github.com/software-mansion-labs/expensify-app-fork/blob/5132dd597e20e4fd79984bb3923b04ff6391e958/src/pages/LogInWithShortLivedAuthTokenPage.tsx#L46)